aip/general/0164.md

```diff
@@ -0,0 +1,155 @@
+  post: "/v1/{name=publishers/*/books/*}:undelete"
+  body: "*"
+}
+```
+
+- The HTTP verb **must** be `POST`.
+- The `body` clause **must** be `"*"`.
+- The response message **must** be the resource itself. There is no
+  `UndeleteBookResponse`.
+  - The response **should** include the fully-populated resource unless it is
+    infeasible to do so.
+  - If the create RPC is [long-running](#long-running-undelete), the response
+    message **must** be a `google.longrunning.Operation` which resolves to the
+    resource itself.
+
+### Undelete request message
+
+Undelete methods implement a common request message pattern:
+
+```proto
+message UndeleteBookRequest {
+  // The name of the deleted book.
+  // Format: publishers/{publisher}/books/{book}
+  string name = 1 [
+    (google.api.field_behavior) = REQUIRED,
+    (google.api.resource_reference).type = "library.googleapis.com/Book"];
+}
+```
+
+- A `name` field **must** be included. It **should** be called `name`.
+  - The field **should** be annotated as required.
+  - The field **should** identify the [resource type][aip-123] that it
+    references.
+  - The comment for the field **should** document the resource pattern.
+- The request message **must not** contain any other required fields, and
+  **should not** contain other optional fields except those described in this
+  or another AIP.
+
+### Long-running undelete
+
+Some resources take longer to undelete a resource than is reasonable for a
+regular API request. In this situation, the API **should** use a long-running
+operation (AIP-151) instead:
+
+```proto
+rpc UndeleteBook(UndeleteBookRequest) returns (google.longrunning.Operation) {
+  option (google.api.http) = {
+    post: "/v1/{name=publishers/*/books/*}:undelete"
+    body: "*"
+  };
+  option (google.longrunning.operation_info) = {
+    response_type: "Book"
+    metadata_type: "OperationMetadata"
+  }
+}
+```
+
+- The response type **must** be set to the resource (what the return type would
+  be if the RPC was not long-running).
+- Both the `response_type` and `metadata_type` fields **must** be specified.
+
+### List and Get
+
+Soft-deleted resources **should not** show up in `List` (AIP-132) requests by
+default (unless `bool show_deleted` is true). `Get` (AIP-131) requests for
+soft-deleted resources **should** return the resource (rather than a
+`NOT_FOUND` error).
+
+APIs that soft delete resources **may** choose a reasonable strategy for
+purging those resources, including automatic purging after a reasonable time
+(such as 30 days), allowing users to set an expiry time (AIP-214), or retaining
+the resources indefinitely. Regardless of what strategy is selected, the API
+**should** document when soft deleted resources will be completely removed.
+
+### Declarative-friendly resources
+
+A resource that is declarative-friendly (AIP-128) **should** support soft
+delete and undelete.
+
+**Important:** There is an ambiguity in declarative tooling between "create"
```

Comment:
After a process of:

- Create XYZ
- Delete XYZ
- Create XYZ

... presumably the "original" XYZ can no longer be undeleted? (I think we discussed this, but I can't remember the result.)

Reply:
The result for now is an error on line 3, with the potential to add a strategy for this later.